src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenSettingsUpdater.cs

```diff
@@ -0,0 +1,77 @@
+using System.Security.Cryptography;
+using System.Threading.Tasks;
+using OrchardCore.Entities;
+using OrchardCore.Environment.Extensions.Features;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Settings;
+using OrchardCore.Modules;
+
+namespace OrchardCore.Media.Processing
+{
+    /// <summary>
+    /// Generates a random key for media token hashing.
+    /// To regenerate the key disabled and enable the feature.
+    /// </summary>
+    public class MediaTokenSettingsUpdater : ModularTenantEvents, IFeatureEventHandler
+    {
+        private readonly ISiteService _siteService;
+
+        public MediaTokenSettingsUpdater(ISiteService siteService)
+        {
+            _siteService = siteService;
+        }
+
+        public override async Task ActivatedAsync()
+        {
+            var mediaTokenSettings = (await _siteService.GetSiteSettingsAsync()).As<MediaTokenSettings>();
+
+            if (mediaTokenSettings.HashKey == null)
+            {
+                var siteSettings = _siteService.LoadSiteSettingsAsync().GetAwaiter().GetResult();
```

Comment:
@jtkech looks good. Haven't tried it yet ;)

Will just need to change these to await (I'll do it shortly, after I try it)

As you say still does not solve the multiple instance issue.

But we will need to have a better solution for that, for migrations as well, so we will remember to apply the same solution here.

Reply:
@deanmarcussen 

> Will just need to change these to await

Oops forgot it

> As you say still does not solve the multiple instance issue.

Not really an issue, that's okay if the `OC.Tenants.Distributed` feature (the one i demoed) is enabled, so that when a node is released / reloaded, other nodes are also released / reloaded

Reply:
Done